crates/bytecode/src/iterator.rs

```diff
@@ -0,0 +1,389 @@
+use crate::{opcode, Bytecode, OpCode};
+
+/// Iterator over opcodes in a bytecode, skipping immediates.
+///
+/// This allows you to iterate through the actual opcodes in the bytecode,
+/// without dealing with the immediate values that follow instructions.
+#[derive(Debug, Clone)]
+pub struct BytecodeIterator<'a> {
+    /// Reference to the underlying bytecode bytes
+    bytes: &'a [u8],
+    /// Current position in the bytecode
+    position: usize,
+    /// End position in the bytecode (to handle original length for legacy bytecode)
+    end: usize,
+}
+
+impl<'a> BytecodeIterator<'a> {
+    /// Creates a new iterator from a bytecode reference.
+    pub fn new(bytecode: &'a Bytecode) -> Self {
+        let bytes = bytecode.bytecode();
+        let end = match bytecode {
+            Bytecode::LegacyAnalyzed(analyzed) => analyzed.original_len(),
+            Bytecode::Eip7702(_) => 0,
+            _ => bytes.len(),
+        };
+
+        Self {
+            bytes: bytes.as_ref(),
+            position: 0,
+            end,
+        }
+    }
+
+    /// Returns the current position in the bytecode.
+    pub fn position(&self) -> usize {
+        self.position
+    }
+
+    /// Skips to the next opcode, taking into account PUSH instructions.
+    pub fn skip_to_next_opcode(&mut self) {
+        if self.position >= self.end {
+            return;
+        }
+
+        let opcode = self.bytes[self.position];
+        self.position += 1;
+
+        if opcode::OpCode::new(opcode).is_none() {
+            // Unknown opcode, return in that case
+            return;
+        }
+
+        // Skip opcode immediates based on immediate size
+        let immediate_size = if opcode == opcode::RJUMPV {
+            if let Some(&max_index) = self.bytes.get(self.position) {
+                // RJUMPV has a variable number of immediates: 1 (max_index) + max_index * 2 bytes for jump targets
+                1 + (max_index as usize) * 2
+            } else {
+                0
+            }
+        } else {
+            opcode::OPCODE_INFO[opcode as usize]
+                .map(|info| info.immediate_size() as usize)
+                .unwrap_or_default()
+        };
```

Comment:
It is nicer code if it is like this
```suggestion
        let immediate_size = opcode::OPCODE_INFO[opcode as usize]
                .map(|info| info.immediate_size() as usize)
                .unwrap_or_default();

        if opcode == opcode::RJUMPV {
            if let Some(max_index) = self.bytes.get(self.position) {
                // RJUMPV has a variable number of immediates: max_index * 2 bytes for jump targets
                immediate_size += (max_index as usize) * 2
            }
        }
```

Decoupling those two